clang/lib/CodeGen/CGExprConstant.cpp

```diff
@@ -2220,6 +2220,11 @@ llvm::Constant *ConstantLValueEmitter::emitPointerAuthPointer(const Expr *E) {
 
   // The assertions here are all checked by Sema.
   assert(Result.Val.isLValue());
+  auto *Base = Result.Val.getLValueBase().get<const ValueDecl *>();
+  if (auto *Decl = dyn_cast_or_null<FunctionDecl>(Base)) {
+    assert(Result.Val.getLValueOffset().isZero());
+    return CGM.getRawFunctionPointer(Decl);
```

Comment:
Why is this code not part of ConstantEmitter()?

Reply:
This function is called to return an unsigned pointer. The returned pointer is used to create a signed pointer for `__builtin_ptrauth_sign_constant` in `emitPointerAuthSignConstant`. `ConstantEmitter` returns a pointer that is signed with zero or a non-zero discriminator that is based on the function type etc.

Reply:
In that case, why are you special-casing functions here?  This looks like emitPointerAuthPointer can return either a signed pointer or an unsigned pointer, depending on what exactly the expression is referring to, which seems bad.

Reply:
The checks in `SemaChecking.cpp` and the assertion in `emitPointerAuthPointer` ensure that the expression of the first argument to `__builtin_ptrauth_sign_constant` evaluates to an `LValue` with a `ValueDecl` base. With that restriction, a call to `emitAbstract` returns a signed pointer only when the argument points to a function definition (see `ConstantLValueEmitter::tryEmitBase`). Since we are special-casing functions here, the call to `emitAbstract` returns an unsigned pointer.

Reply:
Just noticed that there's an unnecessary null check in the code above.

Reply:
Okay, I think I follow what this is doing, then... but it seems like a terrible way to structure the code; it's repeating code, and the relevant logic is split over multiple disconnected functions.

Can we add an argument to ConstantLValueEmitter() so users can request that functions are either signed or unsigned, or something like that?

Reply:
You mean to `ConstantEmitter`?  I'd be okay with a flag that specifically requests a raw pointer; that doesn't seem too error-prone.

Reply:
Revised code looks fine.